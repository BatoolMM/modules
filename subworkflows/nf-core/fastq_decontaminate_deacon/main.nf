include { DEACON_INDEX  } from '../../../modules/nf-core/deacon/index/main'
include { DEACON_FILTER } from '../../../modules/nf-core/deacon/filter/main'

workflow FASTQ_DECONTAMINATE_DEACON {

    take:
    ch_fasta_reads // [ val(meta), [ fasta ], [ reads ] ]

    main:

    ch_versions = Channel.empty()

    ch_fasta = ch_fasta_reads
        .map  { meta, fasta, reads -> [ meta, fasta ] }
    // Check if fastqs are single-end or paired-end and run Deacon accordingly
    ch_reads = ch_fasta_reads
        .map  { meta, fasta, reads ->
            if (meta.single_end) {
                if (reads instanceof List && reads.size() != 1) {
                    error("Error: Check your meta.single_end value. Single-end reads should contain one file only.")
                }
                return [ meta, reads ]
            } else {
                if (!(reads instanceof List) || reads.size() != 2) {
                    error("Error: Check your meta.single_end value. Paired-end reads should contain two files; a forward and a reverse.")
                }
                return [ meta, reads ]
            }
        }

    DEACON_INDEX ( ch_fasta )
    ch_versions = ch_versions.mix(DEACON_INDEX.out.versions.first())

    DEACON_FILTER(DEACON_INDEX.out.index.join(ch_reads))
    ch_versions = ch_versions.mix(DEACON_FILTER.out.versions.first())

    emit:
    index          = DEACON_INDEX.out.index           // channel: [ val(meta), [ index ] ]
    fastq_filtered = DEACON_FILTER.out.fastq_filtered // channel: [ val(meta), [ fastq ] ]
    summary        = DEACON_FILTER.out.log            // channel: [ val(meta), [ log ] ]

    versions = ch_versions                            // channel: [ versions.yml ]
}
