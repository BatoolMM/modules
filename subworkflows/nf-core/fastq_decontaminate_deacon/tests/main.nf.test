nextflow_workflow {

    name "Test Subworkflow FASTQ_DECONTAMINATE_DEACON"
    script "../main.nf"
    workflow "FASTQ_DECONTAMINATE_DEACON"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_decontaminate_deacon"
    tag "deacon"
    tag "deacon/index"
    tag "deacon/filter"

    test("sarscov2 - fastq - single-end") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test', single_end:true ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.index,
                    workflow.out.fastq_filtered,
                    file(workflow.out.summary[0][1]).name,
                    workflow.out.versions.collect { path(it).yaml }
                ).match()}
            )
        }
    }

    test("sarscov2 - fastq - paired-end") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test', single_end:false ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                        ]
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.index,
                    workflow.out.fastq_filtered,
                    file(workflow.out.summary[0][1]).name,
                    workflow.out.versions.collect { path(it).yaml }
                ).match()}
            )
        }
    }

    test("sarscov2 - fastq - single-end - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test', single_end:true ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    workflow.out.versions.collect { path(it).yaml }
                ).match()}
            )
        }
    }
}
