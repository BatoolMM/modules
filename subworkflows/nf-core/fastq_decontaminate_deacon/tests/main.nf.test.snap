{
    "sarscov2 - fastq - single-end - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "genome.idx:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.fq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,721349aff7673fd153a9e391c7eab88b",
                    "versions.yml:md5,f1355ede970f7e54ac452275cfab13d8"
                ],
                "fastq_filtered": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.fq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "index": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "genome.idx:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,721349aff7673fd153a9e391c7eab88b",
                    "versions.yml:md5,f1355ede970f7e54ac452275cfab13d8"
                ]
            },
            [
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_FILTER": {
                        "deacon": "0.5.0"
                    }
                },
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_INDEX": {
                        "deacon": "0.5.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T16:23:17.92702932"
    },
    "sarscov2 - fastq - paired-end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idx:md5,7437b25e6659ebfd5cf9be7b325f90c3"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1.fq:md5,bf1aa22249e7e2b462c6be1a53ac55e9",
                        "test_2.fq:md5,d3106ad7395c78a212e6214177d5c054"
                    ]
                ]
            ],
            "test.json",
            [
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_FILTER": {
                        "deacon": "0.5.0"
                    }
                },
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_INDEX": {
                        "deacon": "0.5.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T16:23:11.422525467"
    },
    "sarscov2 - fastq - single-end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idx:md5,7437b25e6659ebfd5cf9be7b325f90c3"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.fq:md5,2250e99ec61fca0e28e2dcb5124d1a16"
                ]
            ],
            "test.json",
            [
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_FILTER": {
                        "deacon": "0.5.0"
                    }
                },
                {
                    "FASTQ_DECONTAMINATE_DEACON:DEACON_INDEX": {
                        "deacon": "0.5.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T16:23:02.404065827"
    }
}